### PR TITLE
Lightweight Delete In Partition

### PR DESCRIPTION
**Pull request by sunny19930321**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add new Command, Lightweight Delete In Partition
```
DELETE FROM [db.]table [ON CLUSTER cluster] [IN PARTITION partition_expr] WHERE expr;
```
```
VM-114-29-tos :) select * from ads_app_poster_ip_source_channel_di_replicated_local;

SELECT *
FROM ads_app_poster_ip_source_channel_di_replicated_local

Query id: 705f05a4-9d55-4ae4-a6dc-d780d36c36b5

   ┌───imp_date─┬─pgid─┬─ztid─┬─mod_id─┬─biz_id─┬─in_imp_cnt─┬─in_imp_uv─┬─in_clck_cnt─┬─in_clck_uv─┬─el_ip_rtype─┬─el_ip_content_id─┐
   │ 2024-05-29 │ 1    │ 1    │ 1      │ 1      │          1 │         1 │           1 │          1 │ 1           │ 1                │
   └────────────┴──────┴──────┴────────┴────────┴────────────┴───────────┴─────────────┴────────────┴─────────────┴──────────────────┘
   ┌───imp_date─┬─pgid─┬─ztid─┬─mod_id─┬─biz_id─┬─in_imp_cnt─┬─in_imp_uv─┬─in_clck_cnt─┬─in_clck_uv─┬─el_ip_rtype─┬─el_ip_content_id─┐
   │ 2024-05-28 │ 1    │ 1    │ 1      │ 1      │          1 │         1 │           1 │          1 │ 1           │ 1                │
   │ 2024-05-28 │ 1    │ 1    │ 1      │ 1      │          1 │         1 │           1 │          1 │ 1           │ 2                │
   └────────────┴──────┴──────┴────────┴────────┴────────────┴───────────┴─────────────┴────────────┴─────────────┴──────────────────┘

3 rows in set. Elapsed: 0.122 sec. 

VM-114-29-tos :) delete from ads_app_poster_ip_source_channel_di_replicated_local on CLUSTER system_cluster IN PARTITION '2024-05-29' where pgid='1';
 on CLUSTER system_cluster IN PARTITION '2024-05-29' where pgid='1';
DELETE FROM ads_app_poster_ip_source_channel_di_replicated_local ON CLUSTER system_cluster IN PARTITION '2024-05-29' WHERE pgid = '1'

Query id: faacef29-743c-453d-80f0-d035b03e90af

   ┌─host──────┬─port─┬─status─┬─error─┬─num_hosts_remaining─┬─num_hosts_active─┐
   │ 127.0.0.1 │ 9000 │      0 │       │                   0 │                0 │
   └───────────┴──────┴────────┴───────┴─────────────────────┴──────────────────┘

1 row in set. Elapsed: 0.163 sec. 

VM-114-29-tos :) select * from ads_app_poster_ip_source_channel_di_replicated_local;

SELECT *
FROM ads_app_poster_ip_source_channel_di_replicated_local

Query id: 8db02ab4-2f5c-418c-9f34-4018971cf14e

   ┌───imp_date─┬─pgid─┬─ztid─┬─mod_id─┬─biz_id─┬─in_imp_cnt─┬─in_imp_uv─┬─in_clck_cnt─┬─in_clck_uv─┬─el_ip_rtype─┬─el_ip_content_id─┐
   │ 2024-05-28 │ 1    │ 1    │ 1      │ 1      │          1 │         1 │           1 │          1 │ 1           │ 1                │
   │ 2024-05-28 │ 1    │ 1    │ 1      │ 1      │          1 │         1 │           1 │          1 │ 1           │ 2                │
   └────────────┴──────┴──────┴────────┴────────┴────────────┴───────────┴─────────────┴────────────┴─────────────┴──────────────────┘

2 rows in set. Elapsed: 0.009 sec. 

```

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/sql-reference/statements/delete

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by sunny19930321** [comment]
what needs to be improved? @alexey-milovidov

**Comment by alexey-milovidov** [comment]
It was arguing about
> PR Check — More than one changelog category specified: 'New Feature', '### Changelog entry (a 

But now it looks ok.

**Comment by sunny19930321** [comment]
@robot-ch-test-poll1

**Comment by sunny19930321** [comment]
> More than one changelog category specified: 'New Feature', '### Changelog entr

It didn't work out @alexey-milovidov

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 9754620c077b85f5b2cb837416a73f3c831842e2 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67805/9754620c077b85f5b2cb837416a73f3c831842e2/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67805/9754620c077b85f5b2cb837416a73f3c831842e2/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by sunny19930321** [comment]
@alexey-milovidov cc

**Comment by alexey-milovidov** [comment]
Tests are good, but for some reason, packages were not installed in 15 minutes. Waiting for colleagues to take it for review.

**Comment by sunny19930321** [comment]
> Tests are good, but for some reason, packages were not installed in 15 minutes. Waiting for colleagues to take it for review.

OK

File: src/Parsers/ASTDeleteQuery.h [hunk 1]
```diff
@@ -19,6 +19,11 @@ class ASTDeleteQuery : public ASTQueryWithTableAndOutput, public ASTQueryWithOnC
         return removeOnCluster<ASTDeleteQuery>(clone(), params.default_database);
     }
 
+    /** Used in DROP PARTITION, ATTACH PARTITION FROM, UPDATE, DELETE, DELETE FROM queries.
```
Comment (yakov-olkhovskiy) [inline src/Parsers/ASTDeleteQuery.h:22]: This comment makes no much sense - it's plain and simple `DELETE FROM` query - there is no `DROP PARTITION` etc. in this query.

File: src/Interpreters/InterpreterDeleteQuery.cpp [hunk 2]
```diff
@@ -79,6 +79,14 @@ BlockIO InterpreterDeleteQuery::execute()
...
             + (delete_query.cluster.empty() ? "" : " ON CLUSTER " + backQuoteIfNeed(delete_query.cluster))
             + " UPDATE `_row_exists` = 0 WHERE " + serializeAST(*delete_query.predicate);
+        if (delete_query.partition)
+        {
+            alter_query =
+                "ALTER TABLE " + table->getStorageID().getFullTableName()
+                + (delete_query.cluster.empty() ? "" : " ON CLUSTER " + backQuoteIfNeed(delete_query.cluster))
+                + " UPDATE `_row_exists` = 0 IN PARTITION " + serializeAST(*delete_query.partition) + " WHERE "
+                + serializeAST(*delete_query.predicate);
+        }
```
Comment (yakov-olkhovskiy) [inline src/Interpreters/InterpreterDeleteQuery.cpp:89]:
```suggestion
            "ALTER TABLE " + table->getStorageID().getFullTableName()
            + (delete_query.cluster.empty() ? "" : " ON CLUSTER " + backQuoteIfNeed(delete_query.cluster))
            + " UPDATE `_row_exists` = 0"
            + (delete_query.partition ? " IN PARTITION " + serializeAST(*delete_query.partition) : "")
            + " WHERE " + serializeAST(*delete_query.predicate);
```

File: tests/queries/0_stateless/02352_lightweight_delete_in_partition.sql [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+DROP TABLE IF EXISTS t_replicated_merge_tree SYNC;
+
+CREATE TABLE t_merge_tree(time Date, id String , name String) ENGINE = MergeTree() PARTITION BY time ORDER BY id;
+CREATE TABLE t_replicated_merge_tree(time Date, id String, name String) ENGINE = ReplicatedMergeTree('/test/02352/{database}/t_rep','1') PARTITION BY time ORDER BY id;
+
+INSERT INTO t_merge_tree select '2024-08-01', '1', toString(number) FROM numbers(100);
+INSERT INTO t_merge_tree select '2024-08-02', '2', toString(number)  FROM numbers(100);
+
+INSERT INTO t_replicated_merge_tree select '2024-08-01', '1', toString(number) FROM numbers(100);
+INSERT INTO t_replicated_merge_tree select '2024-08-02', '2', toString(number)  FROM numbers(100);
```
Comment (yakov-olkhovskiy) [inline tests/queries/0_stateless/02352_lightweight_delete_in_partition.sql:11]: you need it having no distinction by `id` in order to test what you intended to test
```suggestion
INSERT INTO t_merge_tree select '2024-08-01', '1', toString(number) FROM numbers(100);
INSERT INTO t_merge_tree select '2024-08-02', '1', toString(number)  FROM numbers(100);

INSERT INTO t_replicated_merge_tree select '2024-08-01', '1', toString(number) FROM numbers(100);
INSERT INTO t_replicated_merge_tree select '2024-08-02', '1', toString(number)  FROM numbers(100);
```

File: src/Parsers/ASTDeleteQuery.h (hunk 1 above)
Comment (sunny19930321) [reply src/Parsers/ASTDeleteQuery.h:22]: ok, delete DROP PARTITION, ATTACH PARTITION FROM, UPDATE, DELETE, comment

File: src/Interpreters/InterpreterDeleteQuery.cpp (hunk 2 above)
Comment (sunny19930321) [reply src/Interpreters/InterpreterDeleteQuery.cpp:89]: ok

**Comment by sunny19930321** [comment]
@yakov-olkhovskiy cc

**Review by yakov-olkhovskiy — APPROVED** [review]

**Comment by clickgapai** [comment]
Hi @sunny19930321 @yakov-olkhovskiy — while reviewing this PR I found the following:

- **Bug (P0):** `DELETE ... IN PARTITION` partition clause silently ignored for `EmbeddedRocksDB` and other `supportsDelete()` engines — https://github.com/ClickHouse/ClickHouse/issues/104980

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**